durable-storage/src/merkle_layer/node_resolver.rs

```diff
@@ -0,0 +1,85 @@
+// SPDX-FileCopyrightText: 2025 TriliTech <contact@trili.tech>
+// SPDX-License-Identifier: MIT
+
+use std::sync::Arc;
+
+use octez_riscv_data::serialisation::borrow_decode;
+
+use super::node_wrapper::MavlNodeWrapper;
+use super::node_wrapper::SimpleNodeWrapper;
+use crate::merkle_layer::node::MavlNode;
+use crate::merkle_layer::node::MavlNodeHashRepresentation;
+use crate::merkle_layer::node_wrapper::FullNodeWrapper;
+use crate::persistence_layer::PersistenceLayer;
+
+/// This trait defines the interface for resolving
+/// [`MavlNode`]s.
+pub(crate) trait MavlNodeResolver: Clone {
+    type NodeWrapper: MavlNodeWrapper;
+
+    fn resolve(&self, node_wrapper: &Self::NodeWrapper);
+}
+
+/// The simple resolver assumes everything is in memory
+/// and resolving is a blank operation.
+#[derive(Debug, Default, Clone)]
+pub(crate) struct SimpleMavlNodeResolver {}
+
+impl MavlNodeResolver for SimpleMavlNodeResolver {
+    type NodeWrapper = SimpleNodeWrapper;
+
+    fn resolve(&self, _node_wrapper: &Self::NodeWrapper) {}
+}
+
+/// The full node resolver can resolve nodes
+/// based on a commit hash using the [`PersistenceLayer`].
+/// It can also hold a reference to nodes which are
+/// only present in memory.
+#[derive(Debug, Clone)]
+pub(crate) struct FullMavlNodeResolver {
+    persistence_layer: Arc<PersistenceLayer>,
+}
+
+impl FullMavlNodeResolver {
+    /// Creates and instance of [`MavlNodeResolver`]
+    pub(crate) fn new(persistence_layer: Arc<PersistenceLayer>) -> Self {
+        Self { persistence_layer }
+    }
+}
+
+impl MavlNodeResolver for FullMavlNodeResolver {
+    type NodeWrapper = FullNodeWrapper;
+
+    /// Potentially resolves a [`MavlNode`] from RocksDB
+    fn resolve(&self, node_wrapper: &Self::NodeWrapper) {
+        node_wrapper.node.get_or_init(|| {
+            if let Some(hash) = node_wrapper.commited_hash {
+                let key = hash.as_bytes();
+                let serialized_data = self
+                    .persistence_layer
+                    .blob_get(key)
+                    .expect("The key should be in the database");
```

Comment:
so this function panics if set with a hash that doesn't exist in the db? would be helpful to document this